### DISCUSSION TITLE
Best practice to "select" from cache?

### DISCUSSION THREAD

**Discussion by Kadrian** [urql-graphql/urql#2678] [Q&A] [answered]
Coming from redux, I was used to 
- fetching data somewhere, somehow
- storing & normalizing fetched data in the global state
- using "useSelector" inside components to select partial data from the global state

Would a simple analogy in urql be
- fetch data somewhere with  e.g. `useQuery({ query: UserQuery })`
- use graph cache, data is automatically stored & normalized & merged in the cache
- use `useQuery({ query: UserQuery, { requestPolicy: "cache-only" } })` inside components to select data from the cache?

Or is there some better analogy / best practice around "selecting" already fetched data using urql?

**Comment by JoviDeCroock** [comment, marked as answer]
Currently leveraging `useQuery` is the only way to reach out into the normalized cache from your front-end library point of view.

There are a few gotcha's around this:

- When you have a list of data and you query a single entity we don't automatically link i.e. `todos() --> [Todo:1]` to `todo(id: 1) --> Todo:1` but you can help us do so [by leveraging resolvers](https://formidable.com/open-source/urql/docs/graphcache/local-resolvers/#resolving-entities)
- We have a concept of [partial data](https://formidable.com/open-source/urql/docs/graphcache/schema-awareness/#partial-results) so in the above example if your list view contains all mandatory fields it can resolve a partial entity for you and fetch the rest from your endpoint

**Comment by sarink** [comment]
Coming from redux, it feels weird because you're used to thinking about what's already in the cache. With graphql, your components don't worry about this, they just make queries, and if that data is cached, it comes from the cache, if it's not, it comes from the server 🪄

**Reply by kitten** [reply]
To be fair, I think the mental model I'd recommend here is that GraphQL (not even just the client but just GraphQL) _is_ like the Redux store. In a perfect world GraphQL being a client-server protocol as well is just an implementation detail and we aim to just have a store of data, which is some kind of level of abstraction on top of a GraphQL schema/endpoint.

The selectors are in a way the GraphQL queries. Our schema says what and how we can select data, and we use it to define what data we need and depend on. Of course, in a vast majority of cases that causes an API request to a GraphQL HTTP API, but from a standpoint of the UI, in a perfect world, I think of it just as a definition of data requirements.
